velox/connectors/hive/storage_adapters/s3fs/S3FileSystem.cpp

```diff
@@ -198,6 +207,214 @@ Aws::Utils::Logging::LogLevel inferS3LogLevel(std::string level) {
+// You can upload these object parts independently and in any order.
+/// After all parts of your object are uploaded, Amazon S3 assembles these parts
+/// and creates the object.
+// https://docs.aws.amazon.com/AmazonS3/latest/userguide/mpuoverview.html
+// S3WriteFile uses the Apache Arrow implementation as a reference.
+// https://github.com/apache/arrow/blob/main/cpp/src/arrow/filesystem/s3fs.cc
+// S3WriteFile is not thread-safe.
+// UploadPart is currently synchronous during flush.
+// TODO: Evaluate and add option for asynchronous part uploads.
+// TODO: Implement retry on failure.
+class S3WriteFile::Impl {
+ public:
+  explicit Impl(const std::string& path, Aws::S3::S3Client* client)
+      : client_(client) {
+    getBucketAndKeyFromS3Path(path, bucket_, key_);
+
+    // Check that the object doesn't exist, if it does throw an error.
+    {
+      Aws::S3::Model::HeadObjectRequest request;
+      request.SetBucket(awsString(bucket_));
+      request.SetKey(awsString(key_));
+      auto objectMetadata = client_->HeadObject(request);
+      VELOX_CHECK(!objectMetadata.IsSuccess(), "S3 object already exists");
+    }
+
+    // Create bucket if not present.
+    {
+      Aws::S3::Model::HeadBucketRequest request;
+      request.SetBucket(awsString(bucket_));
+      auto bucketMetadata = client_->HeadBucket(request);
+      if (!bucketMetadata.IsSuccess()) {
+        Aws::S3::Model::CreateBucketRequest request;
+        request.SetBucket(bucket_);
+        auto outcome = client_->CreateBucket(request);
+        VELOX_CHECK_AWS_OUTCOME(
+            outcome, "Failed to create S3 bucket", bucket_, "");
+      }
+    }
+
+    // Initiate the multi-part upload.
+    {
+      Aws::S3::Model::CreateMultipartUploadRequest request;
+      request.SetBucket(awsString(bucket_));
+      request.SetKey(awsString(key_));
+
+      /// If we do not set anything then the SDK will default to application/xml
+      /// which confuses some tools
+      /// (https://github.com/apache/arrow/issues/11934). So we instead default
+      /// to application/octet-stream which is less misleading.
+      request.SetContentType(kApplicationOctetStream);
+
+      auto outcome = client_->CreateMultipartUpload(request);
+      VELOX_CHECK_AWS_OUTCOME(
+          outcome, "Failed initiating multiple part upload", bucket_, key_);
+      uploadState_.id = outcome.GetResult().GetUploadId();
+    }
+
+    // Create a StringStream to store the writes.
+    currentPart_ = Aws::MakeShared<Aws::StringStream>(
+        "PutObjectInputStream",
+        std::stringstream::in | std::stringstream::out |
+            std::stringstream::binary);
+    closed_ = false;
+    fileSize_ = 0;
+  }
+
+  // Appends data to the end of the file.
+  void append(std::string_view data) {
+    VELOX_CHECK(!closed_, "File is closed");
+    // 'flush' API should trigger uploadPart.
+    // But upload part if the maximum part size is reached.
+    if (currentPartSize_ + data.size() > kMaxPartSize) {
```

Comment:
should this be >=

Reply:
`>=` should work according to the documentation. Will do that.